### DISCUSSION TITLE
Compiler error with VS2022

### DISCUSSION THREAD

**Discussion by okerstadt** [timoore/vsgCs#11] [Q&A] [answered]
Hi, I'm trying to compile the code with VS2022 and get a lot of errors:

5>------ Build started: Project: vsgCs, Configuration: Debug x64 ------
5>Styling.cpp
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataArrayView.h(95,24): error C2838: 'Uint8': illegal qualified name in member declaration
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataArrayView.h(95,24): error C2065: 'Uint8': undeclared identifier
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataArrayView.h(101,24): error C2838: 'Uint16': illegal qualified name in member declaration
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataArrayView.h(101,24): error C2065: 'Uint16': undeclared identifier
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataArrayView.h(107,24): error C2838: 'Uint32': illegal qualified name in member declaration
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataArrayView.h(107,24): error C2065: 'Uint32': undeclared identifier
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataArrayView.h(113,24): error C2838: 'Uint64': illegal qualified name in member declaration
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataArrayView.h(113,24): error C2065: 'Uint64': undeclared identifier
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataArrayView.h(95,5): error C2051: case expression not constant
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataArrayView.h(101,5): error C2051: case expression not constant
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataArrayView.h(107,5): error C2051: case expression not constant
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataArrayView.h(113,5): error C2051: case expression not constant
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(30,25): error C2143: syntax error: missing ',' before '*'
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(37,22): error C2143: syntax error: missing ';' before '*'
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(37,79): error C2238: unexpected token(s) preceding ';'
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(144,26): error C2988: unrecognizable template declaration/definition
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(144,26): error C2143: syntax error: missing ',' before '&'
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(240,26): error C2988: unrecognizable template declaration/definition
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(240,26): error C2143: syntax error: missing ',' before '&'
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(312,26): error C2988: unrecognizable template declaration/definition
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(312,26): error C2143: syntax error: missing ',' before '&'
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(346,13): error C2327: 'CesiumGltf::MetadataFeatureTableView::ClassProperty': is not a type name, static, or enumerator
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(346,13): message : see declaration of 'CesiumGltf::MetadataFeatureTableView::ClassProperty'
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(346,26): error C2988: unrecognizable template declaration/definition
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(346,26): error C2143: syntax error: missing ',' before '&'
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(392,13): error C2327: 'CesiumGltf::MetadataFeatureTableView::ClassProperty': is not a type name, static, or enumerator
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(392,13): message : see declaration of 'CesiumGltf::MetadataFeatureTableView::ClassProperty'
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(392,26): error C2143: syntax error: missing ',' before '&'
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(397,13): error C2327: 'CesiumGltf::MetadataFeatureTableView::ClassProperty': is not a type name, static, or enumerator
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(397,13): message : see declaration of 'CesiumGltf::MetadataFeatureTableView::ClassProperty'
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(397,26): error C2988: unrecognizable template declaration/definition
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(397,26): error C2143: syntax error: missing ',' before '&'
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(503,13): error C2327: 'CesiumGltf::MetadataFeatureTableView::ClassProperty': is not a type name, static, or enumerator
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(503,13): message : see declaration of 'CesiumGltf::MetadataFeatureTableView::ClassProperty'
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(503,26): error C2143: syntax error: missing ',' before '&'
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(533,21): error C2143: syntax error: missing ';' before '*'
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(533,37): error C2238: unexpected token(s) preceding ';'
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(534,14): error C2143: syntax error: missing ';' before '*'
5>C:\LibBuild\test\vsgCs\build\_deps\cesium-src\CesiumGltf\include\CesiumGltf/MetadataFeatureTableView.h(534,23): error C2238: unexpected token(s) preceding ';'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(425,19): error C2065: 'Schema': undeclared identifier
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(425,10): error C2923: 'std::optional': 'Schema' is not a valid template type argument for parameter '_Ty'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(425,19): message : see declaration of 'Schema'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(430,43): error C2065: 'Class': undeclared identifier
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(430,16): error C2923: 'std::unordered_map': 'Class' is not a valid template type argument for parameter '_Ty'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(430,43): message : see declaration of 'Class'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(430,16): error C2976: 'std::unordered_map': too few template arguments
5>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\include\unordered_map(67,21): message : see declaration of 'std::unordered_map'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(430,51): error C2955: 'std::unordered_map': use of class template requires template argument list
5>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\include\unordered_map(67,21): message : see declaration of 'std::unordered_map'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(430,59): error C2440: 'initializing': cannot convert from 'std::unordered_map<std::string,CesiumGltf::ExtensionExtFeatureMetadataClass,std::hash<std::string>,std::equal_to<std::string>,std::allocator<std::pair<const std::string,CesiumGltf::ExtensionExtFeatureMetadataClass>>>' to 'const std::unordered_map &'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(430,59): message : Reason: cannot convert from 'std::unordered_map<std::string,CesiumGltf::ExtensionExtFeatureMetadataClass,std::hash<std::string>,std::equal_to<std::string>,std::allocator<std::pair<const std::string,CesiumGltf::ExtensionExtFeatureMetadataClass>>>' to 'const std::unordered_map'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(430,59): message : No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(431,29): error C2039: 'find': is not a member of 'std::unordered_map'
5>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\include\unordered_map(67,21): message : see declaration of 'std::unordered_map'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(432,9): error C3536: 'classItr': cannot be used before it is initialized
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(432,29): error C2039: 'end': is not a member of 'std::unordered_map'
5>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\include\unordered_map(67,21): message : see declaration of 'std::unordered_map'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(436,16): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(436,16): error C2143: syntax error: missing ';' before '&'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(436,18): error C2065: 'defaultClass': undeclared identifier
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(437,43): error C2065: 'ClassProperty': undeclared identifier
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(437,16): error C2923: 'std::unordered_map': 'ClassProperty' is not a valid template type argument for parameter '_Ty'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(437,43): message : see declaration of 'ClassProperty'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(437,16): error C2976: 'std::unordered_map': too few template arguments
5>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\include\unordered_map(67,21): message : see declaration of 'std::unordered_map'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(437,59): error C2955: 'std::unordered_map': use of class template requires template argument list
5>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\include\unordered_map(67,21): message : see declaration of 'std::unordered_map'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(437,72): error C2065: 'defaultClass': undeclared identifier
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(437,70): error C2530: 'properties': references must be initialized
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(438,23): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(438,23): error C2143: syntax error: missing ';' before '&'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(438,25): error C2065: 'featureTable': undeclared identifier
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(439,40): error C2039: 'find': is not a member of 'std::unordered_map'
5>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\include\unordered_map(67,21): message : see declaration of 'std::unordered_map'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(440,9): error C3536: 'classPropertyItr': cannot be used before it is initialized
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(440,40): error C2039: 'end': is not a member of 'std::unordered_map'
5>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\include\unordered_map(67,21): message : see declaration of 'std::unordered_map'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(444,24): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(444,24): error C2143: syntax error: missing ';' before '&'
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(444,26): error C2065: 'property': undeclared identifier
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(445,9): error C2065: 'property': undeclared identifier
5>C:\LibBuild\test\vsgCs\src\vsgCs\Styling.cpp(449,53): error C2065: 'featureTable': undeclared identifier

What can I do to eliminate the errors? Btw: The header file CesiumGltf\include\CesiumGltf/MetadataArrayView.h and some others are removed in actual master of cesium-native.
I have all the needed libraries installed. Is it possible to use these and not the ones downloaded in _deps?

Thanks for your help!
Okerstadt

**Comment by okerstadt** [comment]
As a quick hack i replaced the PropertyType::Uint8... with the corresponding ints and added some include paths. The compiler is happy now but the linker isn't:
12>vsgCsd.lib(vsgResourcePreparer.obj) : error LNK2019: unresolved external symbol "public: __cdecl Cesium3DTilesSelection::TilesetMetadata::~TilesetMetadata(void)" (??1TilesetMetadata@Cesium3DTilesSelection@@QEAA@XZ) referenced in function "public: __cdecl Cesium3DTilesSelection::TileExternalContent::~TileExternalContent(void)" (??1TileExternalContent@Cesium3DTilesSelection@@QEAA@XZ)
13>vsgCsd.lib(vsgResourcePreparer.obj) : error LNK2019: unresolved external symbol "public: __cdecl Cesium3DTilesSelection::TilesetMetadata::~TilesetMetadata(void)" (??1TilesetMetadata@Cesium3DTilesSelection@@QEAA@XZ) referenced in function "public: __cdecl Cesium3DTilesSelection::TileExternalContent::~TileExternalContent(void)" (??1TileExternalContent@Cesium3DTilesSelection@@QEAA@XZ)

**Comment by timoore** [comment, marked as answer]
Hi,
I build with MSVC 2022 and don't see this error. Are you sure there isn't some previous error or file corruption?

The current cesium-native sources have moved to a new glTF extension for presenting metadata to clients, so there's some updating work that needs to be done.

**Comment by okerstadt** [comment]
You are right. I had all libs installed before trying vsgCS and the order of include paths was wrong. So my headers of cesium-native are found and not the ones in _deps. Thanks for the hint.

Now we have to think about switching from osgEarth to vsgCS because the loading of tiles is much faster with vsgCS...
